libbitstream/bitstream.c

```diff
@@ -36,306 +36,279 @@
 
-	return 0;
-}
+	sz = fread(*buf, 1, *len, fp);
+	if (ferror(fp)) {
+		OPAE_ERR("ferror after read");
+		goto out_free;
+	}
 
-STATIC fpga_result string_to_guid(const char *guid, fpga_guid *result)
-{
-	if (uuid_parse(guid, *result) < 0) {
-		OPAE_ERR("Error parsing guid: %s.", guid);
-		return FPGA_INVALID_PARAM;
+	if (sz != *len) {
+		OPAE_ERR("file size and number "
+			 "of bytes read mismatch");
+		goto out_free;
 	}
 
+	fclose(fp);
 	return FPGA_OK;
+
+out_free:
+	free(*buf);
+	*buf = NULL;
+	*len = 0;
+out_close:
+	fclose(fp);
+	return res;
 }
 
-STATIC fpga_result check_bitstream_guid(const uint8_t *bitstream)
+bool opae_is_legacy_bitstream(opae_bitstream_info *info)
 {
-	fpga_guid bitstream_guid;
-	fpga_guid expected_guid;
-	errno_t e;
-
-	e = memcpy_s(bitstream_guid, sizeof(bitstream_guid), bitstream,
-		     sizeof(fpga_guid));
-	if (EOK != e) {
-		OPAE_ERR("memcpy_s failed.");
-		return FPGA_EXCEPTION;
-	}
+	opae_legacy_bitstream_header *hdr;
 
-	/*
-	 * TODO: Would it make sense to encode the metadata
-	 * guid as raw bytes?
-	 */
-	if (string_to_guid(METADATA_GUID, &expected_guid) != FPGA_OK)
-		return FPGA_INVALID_PARAM;
+	if (info->data_len < sizeof(opae_legacy_bitstream_header))
+		return false;
 
-	if (uuid_compare(bitstream_guid, expected_guid) != 0)
-		return FPGA_INVALID_PARAM;
+	hdr = (opae_legacy_bitstream_header *)info->data;
+	if (hdr->legacy_magic == OPAE_LEGACY_BITSTREAM_MAGIC)
+		return true;
 
-	return FPGA_OK;
+	return false;
+}
+
+STATIC void opae_resolve_legacy_bitstream(opae_bitstream_info *info)
+{
+	opae_legacy_bitstream_header *hdr =
+		(opae_legacy_bitstream_header *)info->data;
+	uint8_t *p = &hdr->legacy_pr_ifc_id[15];
+	int i = 0;
+
+	// The guid is encoded backwards.
+	// Reverse it.
+	while (p >= hdr->legacy_pr_ifc_id) {
+		info->pr_interface_id[i++] = *p--;
+	}
+
+	info->rbf_data = info->data + sizeof(opae_legacy_bitstream_header);
+	info->rbf_len = info->data_len - sizeof(opae_legacy_bitstream_header);
 }
 
-STATIC fpga_result get_bitstream_ifc_id(const uint8_t *bitstream,
-					size_t bs_len,
-					fpga_guid *guid)
+STATIC void *opae_bits_parse_metadata(const char *metadata,
+				      fpga_guid pr_interface_id,
+				      int *version)
 {
-	fpga_result result = FPGA_EXCEPTION;
-	char *json_metadata = NULL;
-	uint32_t json_len = 0;
-	const uint8_t *json_metadata_ptr = NULL;
 	json_object *root = NULL;
-	json_object *afu_image = NULL;
-	json_object *interface_id = NULL;
-	errno_t e;
-
-	result = check_bitstream_guid(bitstream);
-	if (result != FPGA_OK)
-		return result;
-
-	json_len = *((uint32_t *)(bitstream + METADATA_GUID_LEN));
-	if (json_len == 0) {
-		OPAE_MSG("Bitstream has no metadata.");
-		result = FPGA_OK;
-		goto out_free;
+	json_object *j_version = NULL;
+	enum json_tokener_error j_err = json_tokener_success;
+	void *parsed = NULL;
+
+        root = json_tokener_parse_verbose(metadata, &j_err);
+	if (!root) {
+		OPAE_ERR("invalid JSON metadata: %s",
+			 json_tokener_error_desc(j_err));
```

Comment:
is json_tokener_success the correct enumerator here?

Reply:
I think so, because the check is on !root.  So, if json_tokener_parse_verbose fails, then root will be NULL and j_err should be populated.